### PR TITLE
fix: Add time slicing to splitstore purging to reduce lock congestion

### PR DESCRIPTION
**Pull request by fridrik01**
## Related Issues
See https://github.com/filecoin-project/lotus/issues/11251

## Context
When splitstore is enabled we perform compaction every 7h30m. During this compaction (which takes several minutes on my fast SSD) there is  relatively small purging step which takes approx 15sec. The purging step chunks the cids which should be purged and operates on them in blocks. In each block it acquires an exclusive lock which it shares with other I/O operation (for example FVM when applying messages). 

So when purging happens, other operations like MpoolSelect block for the whole 15sec duration when it needs to compute tipset state (TipSetState).

## Proposed Changes
I added a simple time slicing to the purging step which only tries to purge for 1 sec, while pausing for 1 sec until it is done. This effectively makes the purge 2x slower, but I don't think we care so much as it only happens every 7h30m and purging is already relatively fast compare the the whole splitstore compaction.

## Test plan
Using the same testing method as described in this [comment](https://github.com/filecoin-project/lotus/issues/11251#issuecomment-1717753704) I observed with this fix that during splitstore compaction the whole TipSetState went down from 15sec to ~2sec.

## Additional Info
See https://github.com/filecoin-project/lotus/issues/11251#issuecomment-1717753704

## Checklist

Before you mark the PR ready for review, please make sure that:

- [x] Commits have a clear commit message.
- [x] PR title is in the form of of `<PR type>: <area>: <change being made>`
  - example: ` fix: mempool: Introduce a cache for valid signatures`
  - `PR type`: fix, feat, build, chore, ci, docs, perf, refactor, revert, style, test
  - `area`, e.g. api, chain, state, market, mempool, multisig, networking, paych, proving, sealing, wallet, deps
- [x] If the PR affects users (e.g., new feature, bug fix, system requirements change), update the CHANGELOG.md and add details to the UNRELEASED section.
- [ ] New features have usage guidelines and / or documentation updates in
  - [ ] [Lotus Documentation](https://lotus.filecoin.io)
  - [ ] [Discussion Tutorials](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] Tests exist for new functionality or change in behavior
- [ ] CI is green

### CODE REVIEW

**Comment by Stebalien** [comment]
We likely either need to:

1. Take some kind of lock when computing tipsets (and/or mining blocks). 1s isn't quite enough to compute a tipset so we'll end up pausing multiple times.
2. Pause longer. IMO, there's no rush here unless I'm missing something.

That's not to say that the current patch won't help. Even if we only pause for 1s, that means we'll be able to spend 50% of the time computing the tipset which means it'll take ~6s on average instead of 3. But I'd still consider pausing for longer.

**Comment by Stebalien** [comment]
> That's not to say that the current patch won't help. Even if we only pause for 1s, that means we'll be able to spend 50% of the time computing the tipset which means it'll take ~6s on average instead of 3. But I'd still consider pausing for longer.

I.e., I'd consider pausing for 4s every 1s, or something like that. That gives us 20% utilization, which means it should only make blocks take 20% longer.

_Ideally_ we'd look at datastore metrics and determine that the datastore is "busy". But that's more complicated.

**Review by Stebalien — APPROVED** [review]

**Comment by Stebalien** [comment]
LGTM but this one needs an approval from @arajasek.

File: blockstore/splitstore/splitstore_compact.go [hunk 1]
```diff
@@ -1372,9 +1372,21 @@ func (s *SplitStore) purge(coldr *ColdSetReader, checkpoint *Checkpoint, markSet
...
 	err := coldr.ForEach(func(c cid.Cid) error {
 		batch = append(batch, c)
 		if len(batch) == batchSize {
+			// add some time slicing to the purge as this a very disk I/O heavy operation that
+			// requires write access to txnLk that may starve other operations that require
+			// access to the blockstore.
+			elapsed := time.Since(now)
+			if elapsed > time.Second {
+				// work 1 second, sleep 4, or 20% utilization
+				time.Sleep(4 * elapsed)
```
Comment (arajasek) [inline blockstore/splitstore/splitstore_compact.go:1386]: `4 * elapsed`? Do we not want `4 * time.Second` here?

**Comment by arajasek** [comment]
I don't see any harm in doing this, but I'm not sure I fully believe the logic here. 

> When splitstore is enabled we perform compaction every 7h30m. During this compaction (which takes several minutes on my fast SSD) there is relatively small purging step which takes approx 15sec. The purging step chunks the cids which should be purged and operates on them in blocks. In each block it acquires an exclusive lock which it shares with other I/O operation (for example FVM when applying messages).

We're saying that the purging step takes 15 seconds, but operates in batches. The batch size is 16k, which seems relatively small to me. We take the lock for each batch.

> So when purging happens, other operations like MpoolSelect block for the whole 15sec duration when it needs to compute tipset state (TipSetState).

In that case, we shouldn't be hogging the lock for the entire 15s, no? We're aggressively trying to grab it between batches, but we should still be releasing the lock for other operations in between batches. Is it the case that a singly batch is taking 15s?

**Review by arajasek — COMMENTED** [review]
See comment about the premise itself. If we're landing, let's make the "work" time and "sleep" time constants (no need to be configurable yet).

File: blockstore/splitstore/splitstore_compact.go (hunk 1 above)
Comment (fridrik01) [reply blockstore/splitstore/splitstore_compact.go:1386]: if we schedule slightly over second, this means that we sleep a bit longer guaranteeing a 20% utilization

Comment (arajasek) [reply blockstore/splitstore/splitstore_compact.go:1386]: Yup, that makes sense.

**Comment by arajasek** [comment]
@fridrik01 Aaah, the lock contention is for every single _message_ application. Got it, that makes complete sense. Thank you!

**Review by arajasek — APPROVED** [review]
Make `time.Second` a constant, but LGTM!

**Pull request merged**